### PR TITLE
API review

### LINKED ISSUE #1941: API Specification
# Spike

## Objective
Research best practices for HTTP APIs (RESTful and/or RPC-over-HTTP). Develop (or use an existing) specification that defines how an HTTP API should be constructed and behave. Examples of questions you should aim to answer are:

- Should endpoint URLs use hyphens or spaces? 
- Some of our endpoints return a 200 code with an error message if there was an error. Is this appropriate behavior? Or should 4xx, 5xx codes be returned instead?
-  How should the different HTTP methods be used? Is it ever appropriate to use a GET to upload data? Or a POST to retrieve data?

## Scope
Time boxed to 3 days.

## Output
- A specification document that defines how our API should behave and be constructed.
- A report that details every way our current API violates the above specification

### PR DESCRIPTION
**Pull request by shreyamalviya**
Fixes #1941

### CODE REVIEW

File: monkey/monkey_island/cc/app.py [hunk 1]
```diff
@@ -170,10 +177,12 @@ def init_api_resources(api: FlaskDIWrapper):
     api.add_resource(PropagationCredentials, "/api/propagation-credentials/<string:guid>")
     api.add_resource(RemoteRun, "/api/remote-monkey")
     api.add_resource(VersionUpdate, "/api/version-update")
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/app.py:179]: Rename to just version or version-info? This doesn't update anything

File: monkey/monkey_island/cc/resources/configuration_import.py [hunk 2]
```diff
@@ -39,6 +39,8 @@ def form_response(self):
 
 
 class ConfigurationImport(flask_restful.Resource):
+    # API Spec: This is an action and there's no "resource"; RPC-style endpoint?
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/resources/configuration_import.py:42]: This should probably be merged with `monkey/monkey_island/cc/resources/island_configuration.py`

File: monkey/monkey_island/cc/resources/exploitations/manual_exploitation.py [hunk 3]
```diff
@@ -7,6 +7,12 @@
 
 
 class ManualExploitation(flask_restful.Resource):
+    # API Spec: RPC-style endpoint? Getting manual exploitations is an action and there's no
+    # "resource", however one could argue that at some point in time, we may want to access
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/resources/exploitations/manual_exploitation.py:11]: I don't understand what qualifies a "resource"? To the API consumer there's no difference if there is a manual exploitation in the database or if it's just created on a get request. This is only used in ransomware report, maybe it should go under the report resource? Like /api/reports/manual-exploitations

File: monkey/monkey_island/cc/resources/island_configuration.py [hunk 4]
```diff
@@ -18,9 +18,14 @@ def get(self):
...
+        # API Spec: Resetting config is an action; separate RPC-style endpoint for this?
         if "reset" in config_json:
             ConfigService.reset_config()
         else:
             if not ConfigService.update_config(config_json, should_encrypt=True):
                 abort(400)
+        # API Spec: We're updating the config and then returning the config back?
+        # RESTfulness of a POST request is to return an identifier of the updated/newly created
+        # resource. Since there's only one thing we're updating (and not multiple "resources"),
+        # should this also be an RPC-style endpoint (/api/resetConfig and /api/updateConfig)?
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/resources/island_configuration.py:30]: We could do a separate get request for the config after the reset

File: monkey/monkey_island/cc/resources/island_mode.py [hunk 5]
```diff
@@ -13,6 +13,9 @@
 
 
 class IslandMode(flask_restful.Resource):
+    # API Spec: Should these be RPC-style endpoints?
+    # POST is not updating/creating a "resource" but setting a property of the Island.
+    # Perhaps /api/setMode and /api/getMode would be more appropriate.
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/resources/island_mode.py:18]: Probably not, I think this should just be an update request

File: monkey/monkey_island/cc/app.py (hunk 1 above)
Comment (mssalvatore) [reply monkey/monkey_island/cc/app.py:184]: Just `version` should be sufficient.

File: monkey/monkey_island/cc/resources/exploitations/manual_exploitation.py (hunk 3 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/resources/exploitations/manual_exploitation.py:11]: I got it slightly mixed up previously. This qualifies as a resource.

File: monkey/monkey_island/cc/resources/blackbox/clear_caches.py [hunk 6]
```diff
@@ -32,6 +34,8 @@ def get(self, **kw):
 
         if ReportService.is_report_generated() or AttackReportService.is_report_generated():
             logger.error(NOT_ALL_REPORTS_DELETED)
+            # API Spec: Why is this 500? 500 should be returned in case an exception occurs on the
+            # server. 40x makes more sense.
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/blackbox/clear_caches.py:38]: 4xx errors mean, "you made a mistake". 5xx errors mean "the server made a mistake".  A 5xx error is appropriate here.

File: monkey/monkey_island/cc/resources/island_configuration.py [hunk 7]
```diff
@@ -18,9 +18,14 @@ def get(self):
     @jwt_required
     def post(self):
         config_json = json.loads(request.data)
+        # API Spec: Resetting config is an action; separate RPC-style endpoint for this?
```
[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/monkey_island/cc/resources/island_configuration.py]: @shreyamalviya I'm not sure why this comment is resolved if you and Vakaris agree there is a resource but the comment is still recommending RPC? Same for a number of other comments.

File: monkey/monkey_island/cc/resources/local_run.py [hunk 8]
```diff
@@ -9,6 +9,8 @@
 from monkey_island.cc.services.run_local_monkey import LocalMonkeyRunService
 
 
+# API Spec: This should be an RPC-style API i.e. two endpoints -
+# "/api/getLocalMonkeyRunningStatus" and "/api/startLocalMonkey"
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/local_run.py]: Camel case or kebab case?

Comment (mssalvatore) [reply monkey/monkey_island/cc/resources/local_run.py]: Getting the status of an agent is REST, not RPC.

File: monkey/monkey_island/cc/resources/monkey_download.py [hunk 9]
```diff
@@ -20,6 +20,7 @@ class UnsupportedOSError(Exception):
 
 
 class MonkeyDownload(flask_restful.Resource):
+    # API Spec: This is an action and there's no "resource"; RPC-style endpoint?
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/monkey_download.py]: The resource is the agent binary.

File: monkey/monkey_island/cc/resources/ransomware_report.py [hunk 10]
```diff
@@ -6,6 +6,8 @@
 
 
 class RansomwareReport(flask_restful.Resource):
+    # API Spec: This is an action and there's no "resource"; RPC-style endpoint?
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/ransomware_report.py]: The report is the resource.

**Review by mssalvatore — CHANGES_REQUESTED** [review]
There are still a lot of comments stating that things are not resources when they are.

File: monkey/monkey_island/cc/app.py [hunk 11]
```diff
@@ -151,11 +156,16 @@ def init_api_resources(api: FlaskDIWrapper):
...
+    # API Spec: Why is the endpoint separated? Why not just
+    # "/api/zero-trust-finding-event/<string:finding_id>"?
     api.add_resource(ZeroTrustFindingEvent, "/api/zero-trust/finding-event/<string:finding_id>")
     api.add_resource(TelemetryFeed, "/api/telemetry-feed")
+    # API Spec: What log? Agent log?
     api.add_resource(Log, "/api/log")
+    # API Spec: Why the inconsistency in endpoints of IslandLog and Log?
     api.add_resource(IslandLog, "/api/log/island/download")
 
+    # API Spec: These two should have the same url syntax
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/app.py:168]: They should probably be the same resource, no? `GET` for download and `POST` for upload.

File: monkey/monkey_island/cc/resources/agent_controls/stop_agent_check.py [hunk 12]
```diff
@@ -4,5 +4,6 @@
 
 
 class StopAgentCheck(flask_restful.Resource):
+    # API Spec: Rename to AgentStopStatus or something
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/resources/agent_controls/stop_agent_check.py:7]: What about `/api/agents/<GUID>/stop-status`?

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**